docs/security/safe-storage-app-secrets.rst

```diff
@@ -0,0 +1,132 @@
+Safe Storage of Application Secrets
+===================================
+
+By `Rick Anderson`_ and `Eilon Lipton`_
+
+This tutorial shows how your application can securely store and access secrets in the local development environment. The most important point is you should never store passwords or other sensitive data in source code, and you shouldn't use production secrets in development and test mode. The secrets manager tool was written to help prevent sensitive data from being checked into source control. The :doc:`../fundamentals/configuration` system that is used by default in DNX based apps can read secrets stored with the secrets manager tool described in this article.
+
+In this article:
+    - `Environment variables`_
+    - `Installing the secrets manager tool`_
+    - `How the secrets manager tool works`_
+    - `Additional Resources`_
+
+Environment variables
+^^^^^^^^^^^^^^^^^^^^^
+
+Your ``Startup`` class should call `AddEnvironmentVariables` as the last configuration method so when 
+`DNX <http://docs.asp.net/en/latest/dnx/overview.html>`_ reads environment variables, and a key is found in a configuration file and the environment, the environment value takes precedence over the configuration file. (See :doc:`../fundamentals/configuration`.) For example, if you create a new ASP.NET web site app with individual user accounts, it will add a default connection string to the *config.json* file. The ``Data:DefaultConnection:ConnectionString`` key value in the *config.json* file uses LocalDB, which runs in user mode and doesn't require a password. When you deploy your application to a test or production server, you can override the ``Data:DefaultConnection:ConnectionString`` key value with an environment variable setting that connects to a test or production SQL Server. That key value would also contain a password to connect to the SQL Server.
+
+Apps frequently require secrets, such as a client ID for OAuth. These passwords and other sensitive data should never be added to *config* files inside your project's source tree, as configuration files can be accidentally checked into source control. The secrets manager tool provides a mechanism to store sensitive data for development work outside your project tree. The secrets manager tool is a DNX console application that is used to store secrets used by DNX and ASP.NET applications in the development environment.

```

Comment:
You’re right. I’ll fix that. Thanks for letting me know.

From: Kévin Chalet [mailto:notifications@github.com]
Sent: Friday, June 26, 2015 2:34 PM
To: aspnet/Docs
Cc: Rick Anderson
Subject: Re: [Docs] Rick anderson secrets (#306)


In docs/security/safe-storage-app-secrets.rst<https://github.com/aspnet/Docs/pull/306#discussion_r33398080>:

> +

> +This tutorial shows how your application can securely store and access secrets in the local development environment. The most important point is you should never store passwords or other sensitive data in source code, and you shouldn't use production secrets in development and test mode. The secrets manager tool was written to help prevent sensitive data from being checked into source control. The :doc:`../fundamentals/configuration` system that is used by default in DNX based apps can read secrets stored with the secrets manager tool described in this article.

> +

> +In this article:

> +    - `Environment variables`_

> +    - `Installing the secrets manager tool`_

> +    - `How the secrets manager tool works`_

> +    - `Additional Resources`_

> +

> +Environment variables

> +^^^^^^^^^^^^^^^^^^^^^

> +

> +Your ``Startup`` class should call `AddEnvironmentVariables` as the last configuration method so when

> +`DNX <http://docs.asp.net/en/latest/dnx/overview.html>`_ reads environment variables, and a key is found in a configuration file and the environment, the environment value takes precedence over the configuration file. (See :doc:`../fundamentals/configuration`.) For example, if you create a new ASP.NET web site app with individual user accounts, it will add a default connection string to the *config.json* file. The ``Data:DefaultConnection:ConnectionString`` key value in the *config.json* file uses LocalDB, which runs in user mode and doesn't require a password. When you deploy your application to a test or production server, you can override the ``Data:DefaultConnection:ConnectionString`` key value with an environment variable setting that connects to a test or production SQL Server. That key value would also contain a password to connect to the SQL Server.

> +

> +Apps frequently require secrets, such as a client ID for OAuth. These passwords and other sensitive data should never be added to *config* files inside your project's source tree, as configuration files can be accidentally checked into source control. The secrets manager tool provides a mechanism to store sensitive data for development work outside your project tree. The secrets manager tool is a DNX console application that is used to store secrets used by DNX and ASP.NET applications in the development environment.

Not important, but the client_id is not a secret thing in OAuth. I guess you meant client_secret?

—
Reply to this email directly or view it on GitHub<https://github.com/aspnet/Docs/pull/306/files#r33398080>.